### PR TITLE
[feat]: [CDS-83146]: Supported InfraExecStep For Remote Yaml.

### PR DESCRIPTION
**Pull request by senjucanon2**
## Describe your changes
Change: Have supported execution for Remote Infrastructures.

Cases Tested: 
- Did a commit on remote infra yaml to create a mismatch in the yaml on remote vs dB. Then in the InfrastructureTaskExecutableStepV2 remote yaml was fetched with the correct gitBranch(same as that of ENV).
-  The gitBranch provided in the yaml for env was used in context for fetching InfraYaml as well.
-  Have added logic for choosing git-branch for execution of Infras. Logic : 
Scenario 1

pipeline -> repo A, branch b1
env (no static linking)
	repo A -> branch b1 picked 
	repo B -> defaultB picked
infra 
	repo A -> branch b1
	repo B -> defaultB
 
Scenario 2

env with branch b2 (static linking)
	infra
		repo A -> b2
		repo B -> defaultB

## Checklist
- [ ] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54800)

### CODE REVIEW

**Comment by senjucanon2** [comment]
trigger uts

**Comment by senjucanon2** [comment]
trigger uts

**Comment by yogesh-chauhan** [comment]
Please add UTs for the same

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54800)* status: 4 file(s) to review. 1 of 2 approvals obtained.

**Comment by mohitg0795** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/infrastructure/InfrastructurePmsPlanCreator.java` line 121 at r2](https://reviewable.io/reviews/harness/harness-core/54800#-Ni4E5I9BxWf0kYMXXeb:-Ni4E5I9BxWf0kYMXXec:b-tuf5bc) ([raw file](https://github.com/harness/harness-core/blob/15853b040ca85532add6ecf33f51e37d64d743d2/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/infrastructure/InfrastructurePmsPlanCreator.java#L121)):*
> ```Java
>     }
>     String gitBranch = null;
>     if (EmptyPredicate.isNotEmpty(environmentYamlV2.getGitBranch())) {
> ```

nit: Instead of having these if statements, we should create a simple utility method to return value if not empty. I'm expecting such methods to already exist in some utils but if its not there, we should definitely create it for better code readability. eg: "getIfNotEmpty" or something similar method name.

**Comment by senjucanon2** [comment]
trigger ut7, ut8, ut9

**Review by senjucanon2 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54800)* status: 3 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/infrastructure/InfrastructurePmsPlanCreator.java` line 121 at r2](https://reviewable.io/reviews/harness/harness-core/54800#-Ni4E5I9BxWf0kYMXXeb:-NiDT6EB1eJzz5Cuviti:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/15853b040ca85532add6ecf33f51e37d64d743d2/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/infrastructure/InfrastructurePmsPlanCreator.java#L121)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

nit: Instead of having these if statements, we should create a simple utility method to return value if not empty. I'm expecting such methods to already exist in some utils but if its not there, we should definitely create it for better code readability. eg: "getIfNotEmpty" or something similar method name.
</blockquote></details>

Done
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2Params.java` line 29 at r2](https://reviewable.io/reviews/harness/harness-core/54800#-Ni4CUlh-pNgwTZKA9Pz:-NiDMfBwCBpnYFIKzz8l:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/15853b040ca85532add6ecf33f51e37d64d743d2/125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2Params.java#L29)):*
<details><summary><i>Previously, hingerabhinav (Abhinav Hinger) wrote…</i></summary><blockquote>

should we call this env git branch so that it is clear we're using env branch for infra?
</blockquote></details>

Done.

**Review by hingerabhinav — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54800)* status: 6 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDCreatorUtils.java` line 87 at r4](https://reviewable.io/reviews/harness/harness-core/54800#-NiJFNd-DWgjR8Kl_WKd:-NiJFNd-DWgjR8Kl_WKe:b-jep3ad) ([raw file](https://github.com/harness/harness-core/blob/e7a85fa1e4028326478ff730db359710bd32b384/125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDCreatorUtils.java#L87)):*
> ```Java
>   }
> 
>   public String getBranchIfNotEmpty(String branch) {
> ```

Instead of here, let's put this in io.harness.ng.core.utils.GitXUtils

**Review by senjucanon2 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54800)* status: 6 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDCreatorUtils.java` line 87 at r4](https://reviewable.io/reviews/harness/harness-core/54800#-NiJFNd-DWgjR8Kl_WKd:-NiJIsYr3_Hz9lMBhG2n:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e7a85fa1e4028326478ff730db359710bd32b384/125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDCreatorUtils.java#L87)):*
<details><summary><i>Previously, hingerabhinav (Abhinav Hinger) wrote…</i></summary><blockquote>

Instead of here, let's put this in io.harness.ng.core.utils.GitXUtils
</blockquote></details>

Done.

**Comment by senjucanon2** [comment]
trigger uts

**Review by hingerabhinav — DISMISSED** [review]
> Reviewed 1 of 1 files at r2, 2 of 5 files at r3, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54800)* status: Some of the checks are too old. 6 file(s) to review. 1 of 2 approvals obtained.

**Comment by senjucanon2** [comment]
trigger checkstylecodeformat

**Review by yogesh-chauhan — DISMISSED** [review]
> Reviewed 1 of 5 files at r3, 4 of 4 files at r5, 1 of 2 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54800)* status: 3 file(s) to review. 1 of 2 approvals obtained.

**Comment by tchaurasiya** [comment]
LGTM, But also lets test for multi infra deployment

**Review by tchaurasiya — APPROVED** [review]

**Comment by senjucanon2** [comment]
trigger uts

File: 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java [hunk 1]
```diff
@@ -426,6 +440,65 @@ && isNotEmpty(stepParameters.getInfraInputs().getValue())) {
...
+      throw new NotFoundException(String.format("Infrastructure with ref [%s] not found", infraRef));
+    }
+    InfrastructureEntity infrastructure = infrastructureEntityOpt.get();
+    if (!StoreType.REMOTE.equals(infrastructure.getStoreType())) {
+      return defaultGitContext;
+    }
+
+    Optional<Environment> environmentOpt = environmentService.getMetadata(AmbianceUtils.getAccountId(ambiance),
+        AmbianceUtils.getOrgIdentifier(ambiance), AmbianceUtils.getProjectIdentifier(ambiance), environmentRef, false);
+    if (!environmentOpt.isPresent()) {
```
Comment (tchaurasiya) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java:463]: nit: can be replaced with isEmpty

**Review by tchaurasiya — APPROVED** [review]

**Comment by tchaurasiya** [comment]
Lets Add the unit test

File: 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java [hunk 2]
```diff
@@ -426,6 +440,65 @@ && isNotEmpty(stepParameters.getInfraInputs().getValue())) {
     return getInfrastructureConfig(infrastructureEntity);
   }
 
+  private GitEntityInfo getGitContextForInfra(
```
Comment (tchaurasiya) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java:443]: lets add the unit test for this logic

Comment (senjucanon2) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/infra/steps/InfrastructureTaskExecutableStepV2.java]: Done , added UT for Static branch linking case

**Comment by senjucanon2** [comment]
trigger uts

**Review by tchaurasiya — APPROVED** [review]

**Review by ankittiwari-harness — APPROVED** [review]

**Pull request merged**